contracts/arbitrage/BancorArbitrage.sol

```diff
@@ -0,0 +1,479 @@
+// SPDX-License-Identifier: SEE LICENSE IN LICENSE
+pragma solidity 0.8.13;
+
+import { ReentrancyGuardUpgradeable } from "@openzeppelin/contracts-upgradeable/security/ReentrancyGuardUpgradeable.sol";
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import { Address } from "@openzeppelin/contracts/utils/Address.sol";
+
+import { IUniswapV2Router02 } from "@uniswap/v2-periphery/contracts/interfaces/IUniswapV2Router02.sol";
+import { ISwapRouter } from "@uniswap/v3-periphery/contracts/interfaces/ISwapRouter.sol";
+import { IWETH } from "@uniswap/v2-periphery/contracts/interfaces/IWETH.sol";
+
+import { Token } from "../token/Token.sol";
+import { TokenLibrary } from "../token/TokenLibrary.sol";
+import { IVersioned } from "../utility/interfaces/IVersioned.sol";
+import { Upgradeable } from "../utility/Upgradeable.sol";
+import { Utils } from "../utility/Utils.sol";
+import { IBancorNetwork, IFlashLoanRecipient } from "../network/interfaces/IBancorNetwork.sol";
+import { PPM_RESOLUTION } from "../utility/Constants.sol";
+import { MathEx } from "../utility/MathEx.sol";
+
+// interface to support Bancor V2 trades
+interface IBancorNetworkV2 {
+    function convertByPath(
+        address[] memory _path,
+        uint256 _amount,
+        uint256 _minReturn,
+        address _beneficiary,
+        address _affiliateAccount,
+        uint256 _affiliateFee
+    ) external payable returns (uint256);
+
+    function conversionPath(Token _sourceToken, Token _targetToken) external view returns (address[] memory);
+}
+
+/**
+ * @dev BancorArbitrage contract
+ */
+contract BancorArbitrage is ReentrancyGuardUpgradeable, Utils, Upgradeable {
+    using SafeERC20 for IERC20;
+    using TokenLibrary for Token;
+    using Address for address payable;
+
+    error InvalidExchangeId();
+    error InvalidRouteLength();
+    error InvalidInitialAndFinalTokens();
+    error InvalidFlashLoanCaller();
+
+    // trade args
+    struct Route {
+        uint16 exchangeId;
+        Token targetToken;
+        uint256 minTargetAmount;
+        uint256 deadline;
+        address customAddress;
+        uint256 customInt;
+    }
+
+    // rewards settings
+    struct Rewards {
+        uint32 percentagePPM;
+        uint256 maxAmount;
```

Comment:
Can probably use `uint224 maxAmount` and reduce this to a single slot.

Reply:
I prefer to keep it at default